app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -2,12 +2,29 @@
 
 module Users
   class OmniauthCallbacksController < Devise::OmniauthCallbacksController
-    # You should configure your model like this:
-    # devise :omniauthable, omniauth_providers: [:twitter]
+    # SAML doesn't support the CSRF protection
+    protect_from_forgery except: %i[samltestid bird]
 
-    # You should also create an action method in this controller like this:
-    # def twitter
-    # end
+    def sso_callback # rubocop:disable Metrics/AbcSize:
+      @user = User.from_omniauth(request.env['omniauth.auth'])
+
+      if @user.persisted?
+        sign_in_and_redirect @user, event: :authentication # this will throw if @user is not activated
+        set_flash_message(:notice, :success, kind: OmniAuth::Utils.camelize(provider)) if is_navigational_format?
+        session['omniauth_provider'] = provider
+      else
+        # Removing extra as it can overflow some session stores
+        session["devise.#{provider}_data"] = request.env['omniauth.auth'].except(:extra)
+        redirect_to new_user_registration_url
+      end
+    end
+
+    alias bird sso_callback
+    alias samltestid sso_callback
```

Comment:
I decided to take an alias and make the method above generic. The documentation "recommends" having a single method for each provider.